sound/soc/raspberrypi/rp1_aout.c

```diff
@@ -0,0 +1,360 @@
+	    params_format(params) != SNDRV_PCM_FORMAT_S16_LE) {
+		dev_err(dai->dev, "Invalid HW params\n");
+		return -EINVAL;
+	}
+
+	audio_mute_sync(aout);
+
+	return 0;
+}
+
+static int rp1_aout_trigger(struct snd_pcm_substream *substream, int cmd,
+			    struct snd_soc_dai *dai)
+{
+	struct rp1_aout *aout = snd_soc_dai_get_drvdata(dai);
+
+	if (cmd == SNDRV_PCM_TRIGGER_STOP ||
+	    cmd == SNDRV_PCM_TRIGGER_SUSPEND ||
+	    cmd == SNDRV_PCM_TRIGGER_PAUSE_PUSH) {
+		/* Push a zero sample (assuming DMA has stopped already?) */
+		aout_reg_wr(aout, AUDIO_OUT_SAMPLE_FIFO, 0);
+	}
+
+	return 0;
+}
+
+static int rp1_aout_dai_probe(struct snd_soc_dai *dai)
+{
+	struct rp1_aout *aout = dev_get_drvdata(dai->dev);
+
+	snd_soc_dai_init_dma_data(dai, &aout->play_dma_data, NULL);
+	snd_soc_dai_set_drvdata(dai, aout);
+
+	return 0;
+}
+
+static const struct snd_soc_dai_ops rp1_aout_dai_ops = {
+	.probe		= rp1_aout_dai_probe,
+	.startup	= rp1_aout_startup,
+	.hw_params	= rp1_aout_hw_params,
+	.trigger	= rp1_aout_trigger,
+};
+
+static const struct snd_soc_component_driver rp1_aout_component = {
+	.name		= "rp1-aout",
+};
+
+static struct snd_soc_dai_driver rp1_aout_dai = {
+	.name		= "rp1-aout",
+	.id		= 0,
+	.playback = {
+		.channels_min = 2,
+		.channels_max = 2,
+		.rates = SNDRV_PCM_RATE_48000,
+		.formats = SNDRV_PCM_FMTBIT_S16_LE,
+	},
+	.ops = &rp1_aout_dai_ops,
+};
+
+static int rp1_aout_platform_probe(struct platform_device *pdev)
+{
+	int ret;
+	struct rp1_aout *aout;
+	struct resource *ioresource;
+
+	aout = devm_kzalloc(&pdev->dev, sizeof(*aout), GFP_KERNEL);
+	if (!aout)
+		return -ENOMEM;
+
+	aout->clk = devm_clk_get(&pdev->dev, NULL);
+	if (IS_ERR(aout->clk))
+		return dev_err_probe(&pdev->dev, PTR_ERR(aout->clk),
+				     "could not get clk\n");
+
+	aout->regs = devm_platform_get_and_ioremap_resource(pdev, 0, &ioresource);
+	if (IS_ERR(aout->regs))
+		return dev_err_probe(&pdev->dev, PTR_ERR(aout->regs),
+				     "could not map registers\n");
+	aout->physaddr = ioresource->start;
+
+	/* Initialize playback DMA parameters */
+	aout->play_dma_data.addr = aout->physaddr + AUDIO_OUT_SAMPLE_FIFO;
+	aout->play_dma_data.addr_width = DMA_SLAVE_BUSWIDTH_4_BYTES;
+	aout->play_dma_data.fifo_size = 16; /* actually 32 but the threshold is 16(?) */
+	aout->play_dma_data.maxburst = 4;
+
+	clk_prepare_enable(aout->clk);
+	aout->dev = &pdev->dev;
+	dev_set_drvdata(&pdev->dev, aout);
+
+	ret = snd_devm_card_new(&pdev->dev, -1, "rp1aout", THIS_MODULE, 0, &aout->card);
+	if (ret)
+		return dev_err_probe(&pdev->dev, ret,
+				     "could not create card\n");
+
+	ret = devm_snd_soc_register_component(&pdev->dev,
+					      &rp1_aout_component,
+					      &rp1_aout_dai, 1);
+	if (ret)
+		return dev_err_probe(&pdev->dev, ret, "failed to register dai\n");
+
+	ret = devm_snd_dmaengine_pcm_register(&pdev->dev, NULL, 0);
+	if (ret)
+		return dev_err_probe(&pdev->dev, ret, "failed to register pcm\n");
+
+	/* TODO: More PM stuff? */
```

Comment:
I don't think this will have an effect unless you define dev_pm_ops with suspend/resume/idle callbacks.

Reply:
Yes, that is what I was wondering. I suppose it ought to be done.

Reply:
Or maybe there's nothing to be done, as there's very little in this component to consume or save power.